feat: i18n for in-browser vscode

Jose-Matsuda opened the pull request:
Will resolve #47 
Still want to test a few things (and then will squash)

Jose-Matsuda commented:
@brendangadd pinging you for a review

brendangadd commented on base/resources/home/.config/Code/User/argv.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "locale":"en"
+}
```
End text files with a newline. _You can set most editors to do this automatically on save._

brendangadd commented on base/resources/home/.config/Code/languagepacks.json:
```diff
@@ -0,0 +1 @@
+{"fr":{"hash":"3143943ae2e0b53c3b4598c110bdfa9e","extensions":[{"extensionIdentifier":{"id":"ms-ceintl.vscode-language-pack-fr","uuid":"MS-CEINTL.vscode-language-pack-fr"},"version":"1.44.0"}],"translations":{"vscode":"/home/jovyan/.vscode/extensions/ms-ceintl.vscode-language-pack-fr-1.44.0/translations/main.i18n.json","vscode.bat":"/home/jovyan/.vscode/extensions/ms-ceintl.vscode-language-pack-f ...
```
Newline

brendangadd commented on base/resources/landing_page/script.js:
```diff
@@ -29,6 +29,18 @@ window.onclick = function (event) {
 
+window.addEventListener("load", (event) => {
+  var xmlHttp = new XMLHttpRequest();
+  const url = window.location.href + "/tools/8081/";
```
Can we use a port less likely to be chosen by the user?

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,40 @@
+
+http
+  .createServer(function (req, res) {
+    res.writeHead(200, { "Content-Type": "text/html" });
```
* We don't know the operation will succeed.
* I would say your response looks more like `text/plain`.

Jose-Matsuda replied:
The bullet point "We don't know the operation will succeed" is the only piece of feedback I don't really understand.

brendangadd replied:
You were writing out a 200 (success) response code right out of the gate, but we don't know, at that point, that the task at hand isn't going to fail. I think you covered this in your latest commits by writing out a 500 for failure cases.

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,40 @@
+http
+  .createServer(function (req, res) {
+    res.writeHead(200, { "Content-Type": "text/html" });
+    var lang = req.headers["accept-language"];
+    lang ? lang : (lang = "en");
+    lang = lang.substring(0, 2);
+
+    lang === "fr" || lang === "en" ? lang : (lang = "en");
```
Clever, and succinct, but the assignment expressions and unused expressions make this difficult to understand. Let's go with something like:
```js
// Some constants at the top
const SUPPORTED_LANGS = [...];
const DEFAULT_LANG = "en";

// Then down here
let lang = (req.headers[...] || DEFAULT_LANG).substring(0, 2);
if (!SUPPORTED_LANGS.includes(lang)) {
  lang = DEFAULT_LANG;
}
```

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,40 @@
+    lang === "fr" || lang === "en" ? lang : (lang = "en");
+
+    // READ THE FILE AND COMPARE TO LANG IF IT CHANGES DO THE FOLLOWING, SYNC to not write to it as well
+    var file = fs.readFileSync(
```
* Don't use `var` unless you mean to. Prefer `const` if the variable will not be reassigned, or `let` if it will.
* Careful on variable names; `file` suggests that the variable holds a file ref rather than its string contents.

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,40 @@
+      "/home/jovyan/.config/Code/User/argv.json",
+      "utf-8"
+    ); // must specify encoding to return string
+    var browserLang = '"locale":"' + lang + '"'; //search using this, "locale":"en" if browserLang not in file then change
+
+    if (file.indexOf(browserLang) == -1) {
+      // if the file index of browserLang == -1 (meaning not found) then write to file
+      fs.writeFile(
+        "/home/jovyan/.config/Code/User/argv.json",
+        '{\n  "locale":"' + lang + '"\n}',
```
You don't want to tackle problems like this with basic string operations if you can help it. It's a pain, difficult to extend in the future, and tends to open the door to injection attacks.
```js
// You can parse the file contents...
const args = JSON.parse(...);
// ... and work with the `locale` key...
if (args.locale !== lang) {
...
// ...  then convert back to a file-friendly string.
JSON.stringify(args);
```

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,40 @@
+
+    if (file.indexOf(browserLang) == -1) {
+      // if the file index of browserLang == -1 (meaning not found) then write to file
+      fs.writeFile(
```
Is there a reason for going with a synchronous read, but an async write?

Jose-Matsuda replied:
My (possibly mis)-understanding was that the synchronous read would prevent a throwing of an error on that line 24 where it would try and retrieve the browserlang before the operation was complete.

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,40 @@
+      fs.writeFile(
+        "/home/jovyan/.config/Code/User/argv.json",
+        '{\n  "locale":"' + lang + '"\n}',
+        function (err) {
+          if (err) throw err;
+          console.log("Updated to current browser language of:" + lang);
+        }
```
Being async, this will throw after the server has already sent a 200 response. I'm also concerned that this uncaught error will crash the server.

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,40 @@
+    ); // must specify encoding to return string
+    var browserLang = '"locale":"' + lang + '"'; //search using this, "locale":"en" if browserLang not in file then change
+
+    if (file.indexOf(browserLang) == -1) {
```
You can often invert a condition like this to switch to a "guard" form. It will limit the amount of code sitting under the condition and make it a bit of an easier read.

```js
if (file.index(browserLang) >= 0) {
  return res.end(...) // The `return` ensures we don't accidentally continue execution.
}

// If we made it here, the language needs to be written.
```

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,40 @@
+      );
+    } // end If not found
+
+    res.end(
+      "This port is used to set the language of vsCode being:(" + lang + " )"
+    );
```
If you're using async calls, this belongs in a callback function. And then you can send an appropriate HTTP response (e.g. 200 vs. 500) depending on the actual result of the operation.

While sync/async doesn't _really_ matter in this case, the async form would definitely be best practice for an HTTP server.

brendangadd commented on base/resources/supervisor/conf.d/browserlang.conf:
```diff
@@ -0,0 +1,7 @@
+autorestart=true
+redirect_stderr=true 
+stdout_logfile=/var/log/supervisor/%(program_name)s.log ; log logs into file
+startretries=5   ; max # of serial start failures (default 3)
```
Newline

brendangadd commented on base/resources/scripts/backup_restore_config.py:
```diff
@@ -56,6 +56,7 @@
     backup_selection = "--include='/.config' \
                         --include='/.config/xfce4/' --include='/.config/xfce4/xfconf/***' \
                         --include='/.config/Code/' --include='/.config/Code/User/' --include='/.config/Code/User/settings.json' \
+                        --include='/.config/Code/User/argv.json' --include='/.config/Code/languagepacks.json'\
```
I don't know what this backup stuff is, but I'm wondering: should everything under `Code/` be included via wildcard pattern?

Jose-Matsuda replied:
I'm also not 100% sure on the backup stuff but I can say that we can include everything under `Code/` without breaking anything. The files that are included are all JSON files that just configure vscode. 
The change will be replacing those two lines with
`--include='/.config/Code/***' \` as rsync's include pattern requires triple asterisk for said directory and everything under it

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,61 @@
+    }
+    const browserLang = '"locale":"' + lang + '"'; //used to search
+
+    function readJSONFile(callback) {
```
For camelcase, use lowercase for trailing characters of an abbreviation: `readJsonFile`

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,61 @@
+      );
+    }
+
+    function createJSON(callback) {
```
Careful on naming. You want, as much as possible, to have names accurately represent what the implementation does, particularly if there's IO involved. If a function does a lot more than it's name suggests, sometimes it's a sign that things can be broken out into chunks.

In this case (`createJson`), it's surprising that the function reads a file, and sometimes also writes a file based on the state of the parent context.

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,61 @@
+
+    function createJSON(callback) {
+      readJSONFile(function detectLangSwitch(returnCode, content) {
+        if (returnCode === 500){
```
Make sure the patterns you use are idiomatic in their respective contexts.
* Node.js likes to pass errors as the first argument to callbacks (null if no error)
* Web layer likes HTTP status codes
So you generally want the top web layer to take errors passed from lower layers and send down appropriate HTTP responses as a result (instead of having lower layers provide HTTP statuses).

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,61 @@
+        if (returnCode === 500){
+          return callback(returnCode, "Error detecting language");
+        }
+        if (content.indexOf(browserLang) >= 0) {
+          return callback(200,"No language switch needed");
+        }
+        //Create JSON to use
+        let JSON_DATA = JSON.parse('{ "locale":"en" }');
+        JSON_DATA.locale = lang;
```
Almost. The idea here is to `JSON.parse` the actual file content instead of searching for a specific string:
* You can mutate its contents (`locale`) and rewrite it, preserving any other config in the file
* Formatting doesn't matter when matching contents as long as the JSON is valid
```json
{"locale":"en"}
// vs.
{"locale":      "en"}
// vs.
{
  "locale": "en"
}
```

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,61 @@
+   used by in-browser VSCode. Uses the accept-language header.
+   Also used by jupyter to determine language.
+ */
+var http = require("http");
```
Make imports `const`

brendangadd commented on base/resources/scripts/languagecheck.js:
```diff
@@ -0,0 +1,61 @@
+   Also used by jupyter to determine language.
+ */
+var http = require("http");
+var fs = require("fs");
+
+http
+  .createServer(function (req, res) {
```
In the future, prefer a promise-based approach, ideally with async/await. This would involve:
```js
const fs = require("fs/promises")
http.createServer(async function(req, res) {
  // Can use await and try/catch in here.
  try {
    await somePromiseFunction();
  } catch (err) {
    // Promise was rejected...
  }
});
```
_Don't rewrite this PR; your callback approach works fine._ :smile:

Commit:
feat:i18n for in-browser vscode
-Added 1.44.0 French language pack by default.
-Added argv.json file with default value of English for vscode.
-Added languagepack.json to have vscode recognize the French language pack.
-English is used if the Accept-Language is not French or English.
-Added onload to landing page to call nodejs
-Added nodejs backend to listen and modify the argv.json.

brendangadd reviewed (approved):
Great commit message. :heart: 
LGTM
